staging/src/k8s.io/apiserver/pkg/storage/value/location_transformer.go

```diff
@@ -0,0 +1,97 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package value
+
+import (
+	"bytes"
+	"fmt"
+)
+
+// LocationTransformer holds a transformer interface and the location of the resources it has to be used for.
+type LocationTransformer struct {
+	Location    string
+	Transformer Transformer
+}
+
+type locationTransformers struct {
+	transformers []LocationTransformer
+	err          error
+}
+
+var _ Transformer = &locationTransformers{}
+
+// NewLocationTransformers supports the Transformer interface by first filtering transformers on
+// whether they are configured to write to the given location. This is done by detecting which transformer's
+// location is a prefix of this resource's location, and choosing the most specific of those locations.
+func NewLocationTransformers(err error, transformers ...LocationTransformer) Transformer {
+	if err == nil {
+		err = fmt.Errorf("the provided value not match any of the supported transformers")
+	}
+	return &locationTransformers{
+		transformers: transformers,
+		err:          err,
+	}
+}
+
+// TransformFromStorage collects all possible transformers (based on their location parameter) and
+// attempts to decrypt using each of them
+func (t *locationTransformers) TransformFromStorage(data []byte, context Context) ([]byte, bool, error) {
+	for _, transformerID := range t.longestMatchingTransformers(context) {
+		result, stale, err := t.transformers[transformerID].Transformer.TransformFromStorage(data, context)
+		if err == nil {
+			return result, stale, err
+		}
+	}
+	return nil, false, t.err
+}
+
+// TransformToStorage finds the first transformer which was configured to write to the location
+// stored in context, and transforms data using that.
+func (t *locationTransformers) TransformToStorage(data []byte, context Context) ([]byte, error) {
+	transformerToUse := t.longestMatchingTransformers(context)
+
+	// If no transformer was assigned to the given location, do not encrypt the data
+	if len(transformerToUse) == 0 {
+		return data, nil
+	}
+
+	// Else use the first matching transformer to write the value to disk.
+	return t.transformers[transformerToUse[0]].Transformer.TransformToStorage(data, context)
+}
+
+// Returns the indices of the transformers which have their location as a prefix of the context, and
+// are the ones with the longest such matching prefix, in a single pass.
+func (t *locationTransformers) longestMatchingTransformers(context Context) []int {
+	var ts []int
+	maxMatch := 0
+	strToMatch := context.AuthenticatedData()
+	if !bytes.HasSuffix(strToMatch, []byte("/")) {
+		strToMatch = append(strToMatch, byte('/'))
+	}
+	for i, transformer := range t.transformers {
+		if bytes.HasPrefix(strToMatch, []byte(transformer.Location)) {
+			length := len(transformer.Location)
+			if length > maxMatch {
+				maxMatch = length
+				ts = []int{i}
+			} else if length == maxMatch {
+				ts = append(ts, i)
```

Comment:
So what you suggest will be correct if you need to do writes. But if you're doing a read, you may need to try out all the transformers which could have written to that location in the past. So for instance, someone configures a transformer1 to write to location A. Then, some days later, the person wants to migrate to transformer2 for location A. There would still be some resources which were written with transformer1, and have not been migrated over yet. Thus we need to return all the transformers which *could* have written to that location, to be able to decrypt the information.

Now there is an issue with my implementation as well, something I had not noticed earlier:
If you initially use transformer1 for writing to location `/a/`, and then add transformer2 for writing to location `/a/b`, when you try to read `/a/b/c`, this function would not return transformer1 in the list. So we shall not be able to decrypt the resource because the TransformFromStorage shall now not use transformer1. I had not noticed this issue earlier.

An approach may be to try all (there won't be many, in most cases) transformers while reading, and only one transformer (what you suggested) while writing. Might have a performance impact, possibly.

Another approach would be to design a good migration scheme which shall detect a change in configuration, and will auto-migrate data to the new configuration. That way we know exactly which transformer was used to write to each location. I suppose this will require persisting the last configuration on disk in plaintext probably, and checking the current configuration against that at cluster bootup (we could, of course, omit the key secrets from this backup).